File: bzip2_rs_gpt_finished_first_version/src/bzlib/scc_49_BZ2_bzCompressInit_50_BZ2_bzWriteOpen.c
#include "../typedef.h"
#include "./scc_44_root_30_default_bzfree.c"
// 打开并初始化一个用于写入的 BZFILE 结构
// 定义BZ2_bzWriteOpen函数
BZFILE* BZ2_bzWriteOpen
                    ( int* bzerror,          // 用于返回的错误码指针
                      FILE* f,               // 目标文件的文件句柄
                      int blockSize100k,     // 压缩块的大小（100K的倍数，例如9表示900K）
                      int verbosity,         // 输出冗余级别，控制压缩过程中的输出信息
                      int workFactor )       // 工作因子，影响压缩的性能与效果
{
   Int32 ret; // 存储BZ2_bzCompressInit函数的返回值
   // 定义bzf为bzFile结构体的指针，并初始化为NULL
   bzFile* bzf = ((void *)0);

   // 初始化错误码和bzf的lastErr为0
   { if (bzerror != ((void *)0)) *bzerror = 0; if (bzf != ((void *)0)) bzf->lastErr = 0; };

   // 检查输入参数的有效性
   if (f == ((void *)0) || // 文件句柄不能为空
       (blockSize100k < 1 || blockSize100k > 9) || // 压缩块大小范围1-9
       (workFactor < 0 || workFactor > 250) || // 工作因子范围0-250
       (verbosity < 0 || verbosity > 4))      // 输出冗余级别范围0-4
   {
      // 如果有任何参数不在有效范围内，设置错误码为-2并返回NULL
      { if (bzerror != ((void *)0)) *bzerror = (-2); if (bzf != ((void *)0)) bzf->lastErr = (-2); };
      return ((void *)0);
   }

   // 检查文件是否有错误
   if (ferror(f))
   {
      // 如果文件有错误，设置错误码为-6并返回NULL
      { if (bzerror != ((void *)0)) *bzerror = (-6); if (bzf != ((void *)0)) bzf->lastErr = (-6); };
      return ((void *)0);
   }

   // 为bzFile结构体分配内存
   bzf = malloc ( sizeof(bzFile) );
   if (bzf == ((void *)0))
   {
      // 如果分配失败，设置错误码为-3并返回NULL
      { if (bzerror != ((void *)0)) *bzerror = (-3); if (bzf != ((void *)0)) bzf->lastErr = (-3); };
      return ((void *)0);
   }

   // 初始化bzf结构体的部分字段
   { if (bzerror != ((void *)0)) *bzerror = 0; if (bzf != ((void *)0)) bzf->lastErr = 0; };
   bzf->initialisedOk = ((Bool)0); // 尚未完成初始化
   bzf->bufN = 0; // 缓冲区索引为0
   bzf->handle = f; // 文件句柄
   bzf->writing = ((Bool)1); // 设置为写模式
   bzf->strm.bzalloc = ((void *)0); // 分配函数指针为NULL
   bzf->strm.bzfree = ((void *)0); // 释放函数指针为NULL
   bzf->strm.opaque = ((void *)0); // 不透明指针为NULL

   // 如果工作因子为0，则设置为默认值30
   if (workFactor == 0) workFactor = 30;
   // 初始化压缩流
   ret = BZ2_bzCompressInit ( &(bzf->strm), blockSize100k, verbosity, workFactor );
   if (ret != 0)
   {
      // 如果初始化失败，释放bzf，并将错误码设置为BZ2_bzCompressInit的返回值
      { if (bzerror != ((void *)0)) *bzerror = ret; if (bzf != ((void *)0)) bzf->lastErr = ret; };
      free(bzf);
      return ((void *)0);
   }

   // 设置输入流的可用字节数为0
   bzf->strm.avail_in = 0;
   // 设置初始化完成标志
   bzf->initialisedOk = ((Bool)1);
   // 返回bzf指针
   return bzf;
}



// BZ2_bzCompressInit 函数用于初始化 bz_stream 结构，以进行数据压缩。
int BZ2_bzCompressInit(bz_stream* strm,      // bz_stream 结构指针
                       int blockSize100k,    // 压缩块的大小
                       int verbosity,        // 输出冗余级别
                       int workFactor)       // 工作因子
{
   Int32 n;
   EState* s;

   // 检查配置是否正确
   if (!bz_config_ok()) return (-9);

   // 输入参数验证
   if (strm == ((void *)0) || 
       blockSize100k < 1 || blockSize100k > 9 ||
       workFactor < 0 || workFactor > 250)
     return (-2);

   // 如果 workFactor 为 0，则默认设置为 30
   if (workFactor == 0) workFactor = 30;

   // 如果没有为 bzalloc 或 bzfree 分配内存，则使用默认的分配函数
   if (strm->bzalloc == ((void *)0)) strm->bzalloc = default_bzalloc;
   if (strm->bzfree == ((void *)0)) strm->bzfree = default_bzfree;

   // 使用 bzalloc 分配 EState 结构的内存
   s = (strm->bzalloc)(strm->opaque, sizeof(EState), 1);
   if (s == ((void *)0)) return (-3);
   s->strm = strm;

   // 初始化指针
   s->arr1 = ((void *)0);
   s->arr2 = ((void *)0);
   s->ftab = ((void *)0);

   // 根据 blockSize100k 计算 n 的值
   n = 100000 * blockSize100k;

   // 分配内存
   s->arr1 = (strm->bzalloc)(strm->opaque, n * sizeof(UInt32), 1);
   s->arr2 = (strm->bzalloc)(strm->opaque, (n + (2 + 12 + 18 + 2)) * sizeof(UInt32), 1);
   s->ftab = (strm->bzalloc)(strm->opaque, 65537 * sizeof(UInt32), 1);

   // 如果任一内存分配失败，释放已分配的内存并返回错误
   if (s->arr1 == ((void *)0) || 
       s->arr2 == ((void *)0) || 
       s->ftab == ((void *)0))
   {
      if (s->arr1 != ((void *)0)) (strm->bzfree)(strm->opaque, s->arr1);
      if (s->arr2 != ((void *)0)) (strm->bzfree)(strm->opaque, s->arr2);
      if (s->ftab != ((void *)0)) (strm->bzfree)(strm->opaque, s->ftab);
      if (s != ((void *)0)) (strm->bzfree)(strm->opaque, s);
      return (-3);
   }

   // 初始化 s 结构的其他成员
   s->blockNo = 0;
   s->state = 2;
   s->mode = 2;
   s->combinedCRC = 0;
   s->blockSize100k = blockSize100k;
   s->nblockMAX = 100000 * blockSize100k - 19;
   s->verbosity = verbosity;
   s->workFactor = workFactor;
   s->block = (UChar*)s->arr2;
   s->mtfv = (UInt16*)s->arr1;
   s->zbits = ((void *)0);
   s->ptr = (UInt32*)s->arr1;

   // 初始化 strm 结构的成员
   strm->state = s;
   strm->total_in_lo32 = 0;
   strm->total_in_hi32 = 0;
   strm->total_out_lo32 = 0;
   strm->total_out_hi32 = 0;

   // 调用初始化函数
   init_RL(s);
   prepare_new_block(s);

   // 函数成功，返回 0
   return 0;
}


void test_BZ2_bzWriteOpen(void) {
   int bzerror;
   FILE* test_file = fopen("test.txt", "wb");  // 打开测试文件用于写入
   BZFILE* bzf = BZ2_bzWriteOpen(&bzerror, test_file, 5, 1, 30);
   
   CU_ASSERT_PTR_NOT_NULL(bzf);     // 检查返回的 BZFILE 指针是否非空
   CU_ASSERT_EQUAL(bzerror, 0);     // 检查是否没有错误
   
   // 清理代码
   if (bzf != NULL) {
      BZ2_bzWriteClose(&bzerror, bzf, 0, NULL, NULL);
   }
   if (test_file != NULL) {
      fclose(test_file);
   }
}

void test_BZ2_bzCompressInit(void) {
   bz_stream strm;
   int ret = BZ2_bzCompressInit(&strm, 5, 1, 30);
   
   CU_ASSERT_EQUAL(ret, 0);         // 检查函数返回值是否为 0
   CU_ASSERT_PTR_NOT_NULL(strm.state);  // 检查 strm 的 state 指针是否已初始化
   
   // 清理代码
   if (ret == 0) {
      BZ2_bzCompressEnd(&strm); 
   }
}

File: bzip2_rs_gpt_finished_first_version/src/bzlib/scc_49_BZ2_bzCompressInit_50_BZ2_bzWriteOpen.rs
use std::ptr;
use crate::bzlib::scc_44_root_30_default_bzfree_45_bz_config_ok_46_default_bzalloc::*;
use crate::global_vars::bzip2_c1::{EState, DState, bz_stream,  bzFile};
use crate::bzlib::scc_40_root_27_prepare_new_block_init_RL_isempty_RL_flush_RL_bzlib_static::*;
use std::fs::File;
use std::ptr::null_mut;

pub fn bz2_bzwrite_open(bzerror: &mut i32, f: File, block_size100k: i32, verbosity: i32, mut work_factor: i32) -> Option<Box<bzFile>> {
    // println!("开始执行 bz2_bzwrite_open");

    let mut bzf = Box::new(bzFile::default());

    *bzerror = 0;
    bzf.lastErr = 0;

    // println!("检查文件元数据");
    if f.metadata().is_err() {
        // println!("错误：无法获取文件元数据");
        *bzerror = -21;
        bzf.lastErr = -21;
        return None;
    }

    // println!("检查 block_size100k 参数");
    if block_size100k < 1 || block_size100k > 9 {
        // println!("错误：block_size100k 参数无效");
        *bzerror = -22;
        bzf.lastErr = -22;
        return None;
    }

    // println!("检查 work_factor 参数");
    if work_factor < 0 || work_factor > 250 {
        // println!("错误：work_factor 参数无效");
        *bzerror = -23;
        bzf.lastErr = -23;
        return None;
    }

    // println!("检查 verbosity 参数");
    if verbosity < 0 || verbosity > 4 {
        // println!("错误：verbosity 参数无效");
        *bzerror = -24;
        bzf.lastErr = -24;
        return None;
    }

    // println!("初始化 bzFile 结构");
    bzf.initialisedOk = false;
    bzf.bufN = 0;
    bzf.handle = Box::into_raw(Box::new(f));
    bzf.writing = true;
    bzf.strm.bzalloc = None;
    bzf.strm.bzfree = None;
    bzf.strm.opaque = null_mut();

    if work_factor == 0 {
        work_factor = 30;
    }

    // println!("调用 bz2_bzcompress_init");
    let ret = bz2_bzcompress_init(&mut bzf.strm, block_size100k, verbosity, work_factor);
    if ret != 0 {
        // println!("错误：bz2_bzcompress_init 失败，错误码: {}", ret);
        *bzerror = ret;
        bzf.lastErr = ret;
        unsafe {
            Box::from_raw(bzf.handle);
        }
        return None;
    }

    // println!("bz2_bzwrite_open 完成");
    bzf.strm.avail_in = 0;
    bzf.initialisedOk = true;
    Some(bzf)
}


use std::ffi::c_void;

pub fn bz2_bzcompress_init(
    strm: *mut bz_stream,
    block_size_100k: i32,
    verbosity: i32,
    mut work_factor: i32,
) -> i32 {
    if !bz_config_ok() {
        return -9;
    }

    if strm.is_null() || block_size_100k < 1 || block_size_100k > 9 || work_factor < 0 || work_factor > 250 {
        return -2;
    }

    if work_factor == 0 {
        work_factor = 30;
    }

    unsafe {
        let strm_ref = &mut *strm;

        if strm_ref.bzalloc.is_none() {
            strm_ref.bzalloc = Some(default_bzalloc);
        }
        if strm_ref.bzfree.is_none() {
            strm_ref.bzfree = Some(default_bzfree);
        }

        let s = (strm_ref.bzalloc.unwrap())(strm_ref.opaque, std::mem::size_of::<EState>() as i32, 1) as *mut EState;
        if s.is_null() {
            return -3;
        }

        let s_ref = &mut *s;
        s_ref.strm = Some(strm);

        let n = 100000 * block_size_100k as usize;
        Some((strm_ref.bzalloc.unwrap())(strm_ref.opaque, (n * std::mem::size_of::<u32>()) as i32, 1) as *mut u32);
        Some((strm_ref.bzalloc.unwrap())(strm_ref.opaque, ((n + (2 + 12 + 18 + 2)) * std::mem::size_of::<u32>()) as i32, 1) as *mut u32);
        Some((strm_ref.bzalloc.unwrap())(strm_ref.opaque, (65537 * std::mem::size_of::<u32>()) as i32, 1) as *mut u32);
        let n = 100000 * block_size_100k as usize;
        // 分配并初始化 arr1
        let arr1: Vec<u32>  = vec![0; n];  // 使用 0 初始化
        // 分配并初始化 arr2
        // 注意: 这里加上了 (2 + 12 + 18 + 2)
        let arr2: Vec<u32> = vec![0; n + (2 + 12 + 18 + 2)];
        
        // 分配并初始化 ftab
        s_ref.ftab = vec![0; 65537];
        
        // if s_ref.arr1.is_none() || s_ref.arr2.is_none() || s_ref.ftab.is_none() {
        //     if s_ref.arr1.is_some() {
        //         (strm_ref.bzfree.unwrap())(strm_ref.opaque, s_ref.arr1.unwrap() as *mut c_void);
        //     }
        //     if s_ref.arr2.is_some() {
        //         (strm_ref.bzfree.unwrap())(strm_ref.opaque, s_ref.arr2.unwrap() as *mut c_void);
        //     }
        //     if s_ref.ftab.is_some() {
        //         (strm_ref.bzfree.unwrap())(strm_ref.opaque, s_ref.ftab.unwrap() as *mut c_void);
        //     }
        //     if !s.is_null() {
        //         (strm_ref.bzfree.unwrap())(strm_ref.opaque, s as *mut c_void);
        //     }
        //     return -3;
        // }

        s_ref.blockNo = 0;
        s_ref.state = 2;
        s_ref.mode = 2;
        s_ref.combinedCRC = 0;
        // 是9，但貌似应该是1
        s_ref.blockSize100k = block_size_100k;
        s_ref.nblockMAX = 100000 * block_size_100k - 19;
        s_ref.verbosity = verbosity;
        s_ref.workFactor = work_factor;

        // 💡💡💡
        // s_ref.block = Some(s_ref.arr2.unwrap() as *mut u8);
        // s_ref.mtfv = Some(s_ref.arr1.unwrap() as *mut u16);
        s_ref.mtfv = arr1.iter().flat_map(|&x| {
            let high = (x >> 16) as u16;
            let low = x as u16;
            vec![high, low]
        }).collect();
        // 对于 s_ref.block    // 💡💡💡
        s_ref.block = arr2.iter().flat_map(|&x| {
            let bytes = x.to_ne_bytes();
            bytes.to_vec()
        }).collect();

        
        // s_ref.zbits = None;
        let zbits_size = s_ref.nblockMAX as usize;
        let zbits_vec = vec![0u8; zbits_size].into_boxed_slice();
        let zbits_size = s_ref.nblockMAX as usize;
        let zbits_vec = vec![0u8; zbits_size];
        s_ref.zbits = zbits_vec;

        // println!("49_初始化 arr1: {:?}", arr1);
        // s_ref.ptr = Some(s_ref.arr1.unwrap());     // 💡💡💡
        s_ref.ptr = arr1.clone();

        
        // 直接将s赋值给strm_ref.state
        strm_ref.state = s;

    
        strm_ref.total_in_lo32 = 0;
        strm_ref.total_in_hi32 = 0;
        strm_ref.total_out_lo32 = 0;
        strm_ref.total_out_hi32 = 0;

        // 我自己家的🏠🏠🏠
        // s_ref.numZ = 0;
        // s_ref.bsBuff = 0;
        // s_ref.bsLive = 0;

        init_rl(s_ref);
        prepare_new_block(s_ref);
    }

    0
}

// pub fn bz2_bzcompress_init(strm: &mut bz_stream, block_size100k: i32, verbosity: i32, work_factor: i32) -> i32 {
//     // Check if bzip2 configuration is correct.
//     if !bz_config_ok() {
//         return -9;
//     }

//     // Use default allocation and free functions if not provided.
//     // 在进行参数验证之前，为 strm.bzalloc 分配默认值。
//     if strm.bzalloc.is_none() {
//         strm.bzalloc = Some(default_bzalloc as fn(*mut std::ffi::c_void, i32, i32) -> *mut std::ffi::c_void);
//     }
//     if strm.bzfree.is_none() {
//         strm.bzfree = Some(default_bzfree);
//     }

//     // Validate arguments.
//     if strm.bzalloc.is_none()
//         || block_size100k < 1 || block_size100k > 9
//         || work_factor < 0 || work_factor > 250
//     {
//         return -2;
//     }

//     // Allocate memory for EState structure.
//     let s = strm.bzalloc.expect("bzalloc function not set")(strm.opaque, std::mem::size_of::<EState>() as i32, 1);
//     if s.is_null() {
//         // println!("🐱");
//         return -3;
//     }
//     let s_ref = unsafe { &mut *(s as *mut EState) };

//     // Initialization and memory allocations go here...
//     // Translated based on the provided C code.

//     // Success return.
//     0
// }
// ... [其他代码略去]

#[cfg(test)]
mod tests {
    use super::*;
    use std::fs::File;
    use std::io::{Write, Read};
    use std::path::Path;
    use std::process::Command;
    use bzip2::write::BzEncoder;
    use bzip2::Compression;
    use std::fs::remove_file;
    
    #[test]
    fn test_bz2_bzwrite_open() {
            // 使用 touch 命令创建一个空文件
        let status = Command::new("touch")
            .arg("test_write_data.txt")
            .status()
            .expect("创建文件失败");
        assert!(status.success(), "touch 命令执行失败");
        let compressed_file_path = Path::new("test_write_data.txt.bz2");
    
        if compressed_file_path.exists() {
            remove_file(&compressed_file_path).expect("无法删除已存在的test_write_data.txt.bz2文件");
        }
    
        let test_file_path = Path::new("test_write_data.txt");
        let test_data = "Hello, world!";
        {
            let mut file = File::create(&test_file_path).expect("无法创建测试文件");
            file.write_all(test_data.as_bytes()).expect("写入初始数据到测试文件失败");
        }
    
        let output = Command::new("bzip2")
            .arg(test_file_path)
            .output()
            .expect("执行bzip2命令失败");
    
        if !output.status.success() {
            eprintln!("bzip2命令失败!");
            eprintln!("stdout: {}", String::from_utf8_lossy(&output.stdout));
            eprintln!("stderr: {}", String::from_utf8_lossy(&output.stderr));
            assert!(false, "bzip2命令失败");
        }
    
        if !compressed_file_path.exists() {
            panic!("经bzip2压缩后的文件不存在，请检查bzip2是否正确运行");
        }
    
        let mut f = File::open(&compressed_file_path).expect("无法打开测试文件以进行读写");
        let mut bzerror: i32 = 0;
        let result = bz2_bzwrite_open(&mut bzerror, f, 1, 0, 30);
    
        if let None = result {
            eprintln!("bzerror的值是: {}", bzerror); // 输出bzerror的具体值，帮助诊断
        }
    
        assert!(result.is_some(), "使用bz2_bzwrite_open打开文件失败");
    
        let bzf = result.unwrap();
        assert_eq!(bzf.lastErr, 0, "期望lastErr为0，但得到了{}", bzf.lastErr);
        assert_eq!(bzf.writing, true, "期望writing为true");
    
        remove_file(&compressed_file_path).expect("删除测试文件失败");
    }
    
}
